### PR TITLE
Rewrite retry mechanism

### PR DESCRIPTION
**Pull request by mickeylee25**
The original approach is very good but I think it's too convoluted to create a class `DeviceRegistrationLoopHandler` to handle the retry mechanism with `DispatchSemaphore`. It also caused a flaky test when testing the retry.

Since we already have a worker, it's convenient to use `retryWhen` operator of RxSwift with serial dispatch queue to do the retry. The retry interval is handled by the `Observable<Int>.timer` and is injected to the worker.

### CODE REVIEW

**Comment by mickeylee25** [comment]
The retry result with Charles rewrite
[](url)

File: DummyProject/DummyProject/AppDelegate.swift [hunk 1]
```diff
@@ -16,7 +17,8 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
                      didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
         let configuration = SDKConfiguration(
             appCode: "LS",
-            clientSecret: "$2y$10$O7HK3Afr1PZH3WTiQ7bTg.kfcle88e/n9GqrcCp7qWH8Rvv.Ojl/C")
```
Comment (phchosis) [inline DummyProject/DummyProject/AppDelegate.swift:19]: We should find a way to scrub this from the repo as I don't think we should be committing secrets.

File: General/DeviceRegistration/DeviceRegistrationWorker.swift [hunk 2]
```diff
@@ -9,59 +9,86 @@
...
     }
 
     init(
         staticDeviceInformation: StaticDeviceInformation,
         reachabilityChecker: ReachabilityChecking,
-        loopHandler: DeviceRegistrationLoopHandling,
-        deviceRegisteredSubject: CurrentValueSubject<Bool, Never>,
-        store: Storeable
+        deviceProvider: DeviceProviding.Type = DeviceRepository.self,
+        deviceRegistrationMaxAttempts: Int = .max,
```
Comment (phchosis) [inline General/DeviceRegistration/DeviceRegistrationWorker.swift:47]: could using .max mean that it'll try to register this many times: [9223372036854775807](https://www.tutorialspoint.com/swift/swift_data_types.htm)?

**Review by phchosis — APPROVED** [review]
just the question around .max but looks good

File: General/DeviceRegistration/DeviceRegistrationWorker.swift (hunk 2 above)
Comment (mickeylee25) [reply General/DeviceRegistration/DeviceRegistrationWorker.swift:47]: Yes, using .max keeps retrying until the user quits the app. I think it's risky to set the default value as max but couldn't find any suggested retry time on the mobile SDK doc. Assuming the retry interval is 10 seconds by default, setting the maximum retry attempts to 60 times takes 10 minutes to see the final error.

How do you feel about we change the default value to 60 or 360 attempts? It takes 10 minutes or 1 hour to retry the last time. It makes more sense than .max in my opinion as no users will wait that long

File: DummyProject/DummyProject/AppDelegate.swift (hunk 1 above)
Comment (mickeylee25) [reply DummyProject/DummyProject/AppDelegate.swift:19]: Yeah, totally agree with you. I think we can approach it with below options at the moment.
Option 1. We can leave fake appCode, clientSecret and apiUrl here and let QA fills these up by themselves in codes.
Option 2. Create a plist file which contains appCode, clientSecret and apiUrl.

The option 2 is a bit redundant as the developers still can access the plist file via the Xcode project, so I'd prefer option 1, but we have to make sure QA knows how to change these configurations in codes.

Comment (phchosis) [reply DummyProject/DummyProject/AppDelegate.swift:19]: Yeah, Option 1 sounds better, we can write a guide for them on confluence

Comment (mickeylee25) [reply DummyProject/DummyProject/AppDelegate.swift:19]: Thanks. I removed the configuration details from the repo and attach the tutorial to the ticket.

**Pull request merged**